File: core/src/main/scala/tripaint/effects/LocalEffect.scala
package tripaint.effects

import tripaint.{Color, ColorLookup, FloodFillSearcher}
import tripaint.coords.{GlobalPixCoords, GridCoords, PixelCoords}
import tripaint.grid.{ImageGrid, ImageGridColorLookup}

abstract class LocalEffect extends Effect {

  protected def predicate(image: ColorLookup, here: GlobalPixCoords)(
      coords: GlobalPixCoords,
      color: Color
  ): Boolean

  protected def weightedColor(image: ColorLookup, here: GlobalPixCoords)(
      coords: GlobalPixCoords
  ): (Double, Color)

  override def action(images: Seq[GridCoords], grid: ImageGrid): Unit = {
    val colorLookup = new ImageGridColorLookup(grid)

    val searcher = new FloodFillSearcher(colorLookup)
    val allChanges = for (imageCoords <- images) yield {
      val image = grid(imageCoords).get.storage

      val newVals = for (here <- image.allPixels) yield {
        val coords = PixelCoords(here, imageCoords)
        val coordsGlobal = coords.toGlobal(grid.imageSize)

        val cols = searcher
          .search(coordsGlobal, predicate(colorLookup, coordsGlobal))
          .map(weightedColor(colorLookup, coordsGlobal))

        val numCols = cols.foldLeft(0d)(_ + _._1)
        here -> (cols.map((w, c) => c * w).reduce(_ + _) / numCols)
      }
      imageCoords -> newVals
    }
    for ((im, vals) <- allChanges) {
      val image = grid(im).get.storage
      for ((coords, color) <- vals) image.setColor(coords, color)
    }
  }
}

File: core/src/main/kotlin/tripaint/effects/LocalEffect.kt
package tripaint.effects

import tripaint.ColorLookup
import tripaint.FloodFillSearcher
import tripaint.color.Color
import tripaint.coords.GlobalPixCoords
import tripaint.coords.GridCoords
import tripaint.coords.PixelCoords
import tripaint.grid.ImageGrid
import tripaint.grid.ImageGridColorLookup

abstract class LocalEffect : Effect {

    protected abstract fun predicate(image: ColorLookup, here: GlobalPixCoords): (coords: GlobalPixCoords, color: Color) -> Boolean

    protected abstract fun weightedColor(image: ColorLookup, here: GlobalPixCoords): (coords: GlobalPixCoords) -> Pair<Double, Color>

    override fun action(images: List<GridCoords>, grid: ImageGrid) {
        val colorLookup = ImageGridColorLookup(grid)

        val searcher = FloodFillSearcher(colorLookup)
        val allChanges = images.map { imageCoords ->
            val image = grid.apply(imageCoords)!!.storage

            val newVals = image.allPixels().map { here ->
                val coords = PixelCoords.from(here, imageCoords)
                val coordsGlobal = coords.toGlobal(grid.imageSize)

                val cols = searcher
                    .search(coordsGlobal, predicate(colorLookup, coordsGlobal))
                    .map(weightedColor(colorLookup, coordsGlobal))

                val numCols = cols.fold(0.0) { acc, b -> acc + b.first }
                Pair(here, (cols.map { (w, c) -> c * w }.reduce { a, b -> a + b } / numCols))
            }
            Pair(imageCoords, newVals)
        }
        for ((im, vals) in allChanges) {
            val image = grid.apply(im)!!.storage
            for ((coords, color) in vals) {
                image.setColor(coords, color)
            }
        }
    }
}
